src/blocked.c

```diff
@@ -278,62 +278,48 @@ void serveClientsBlockedOnListKey(robj *o, readyList *rl) {
                 continue;
             }
 
+            /* fixme There is a situation:
+             * blmpop in serveClientBlockedOnList -> listPopRangeAndReplyWithKey
+             * -> listElementsRemoved -> listTypeLength
+             *
+             * listTypeLength(o) == 0, it will call `dbDelete` delete the empty list
+             * and back in here, there will be a extra listTypeLength call.
+             * In cluster mode, there will be a problem since `o` became a string type.
+             *
+             * In here `if (server.cluster_enabled) slotToKeyDel(key->ptr);`
+             * Don't know why, don't know rax too much...
+             *
+             * That will happen in test-external-cluster: `BLMPOP with multiple blocked clients` test case.
+             *
+             * This is the reason why i added a `del = 0` in blocking way before.
+             * Avoid calling listTypeLength after deleting the key (or delete the key in caller)
+             * */
+            //if (listTypeLength(o) == 0) {
+            if (!isListType(o) || listTypeLength(o) == 0) {
+                /* The list is empty, we can't pop any elements, break the loop. */
+                break;
+            }
```

Comment:
not sure i understand everything, but i'll try to respond, and you'l tell me if there's anything still missing or not:
1. isn't hte call to `isListType` causing an invalid read? (valgrind will complain). i.e. wasn't `o` already released?
2. i think the right clean solution for that is that `serveClientBlockedOnList` returns an indication that the key was deleted, and then you can break out of the loop (instead of checking if the list is empty or is no longer a list).
3. i'm not sure what did you mean about cluster node and `o` becoming a string type, maybe that's just caused by the fact `o` was released and this memory is now used by something else?
4. not sure what you meant about `slotToKeyDel` can you provide more info or a reference to where it is called? or maybe it's also related to the fact the key was released (missing an incrRefcount)?

i haven't yet completed a review on the rest of the changes in this commit, but i'll post this one first so you can respond sooner.

Reply:
> i haven't yet completed a review on the rest of the changes in this commit, but i'll post this one first so you can respond sooner.

That was nice. Thank you and sorry for my fuzzy statement...
My question was: looking at my last commit that can pass other CI tests but it will fail in the external-cluster.
So if `o` has been released, the next `listTypeLength(o)` should also fail in normal test.
I am worried that it is because of the problem that I use, or the original code is wrong in someway

Reply:
i'm not certain i understand your question, but if you're concerned about the fact that only the cluster tests fail, and the others don't, i suggest you try to run that test with valgrind (`make distclean; make valgrind; ./runtest --single unit/type/list-common --valgrind`) and you'll find out that there are serious issues even without cluster mode.

Reply:
1. i don't understand why would `o` become a string type. i think it should have been released and you're just reading random memory.
2. i think this is the right solution, i.e. instead of telling the code inside list to delete or not to delete, it should always do the deletion (together with all the other nuances of the deletion, like notification etc), and just return an indication to the caller that the object passed as reference is no longer valid (was deleted / released)
3. see 1
4. who's calling dbSyncDelete in that case? is the crash happening because `key->ptr` is no longer valid? or because the cluster slots data structure was compromised (or doesn't support double deletion)? i suggest to look into that with valgrind to gain a better insight.

Reply:
> you'll find out that there are serious issues even without cluster mode.

Yes you are right. A little embarrassing.. Thanks for the answer, i will try to work it out